Verify that user uploaded an image before storing BlobKey

anna-giang opened the pull request:
Ensure that user is uploading only files of type .jpeg, .jpg or .png (typical image files). 
If the file uploaded is not one of these 3 types, delete the Blob of the uploaded file, and return none. Otherwise return the BlobKey of the stored image file.

anna-giang committed:
Verify that user uploaded an image before storing BlobKey

anna-giang committed:
Store image type strings in final class member

The permitted upload filetypes were previously hardcoded in the logic of the  getBlobKey function, which is not best practice.

anna-giang commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -41,6 +45,9 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  // Supported upload file types of the comment form
+  private final HashSet<String> FILETYPES = new HashSet<String>(
```
Is this better practice than defining the HashSet directly in the getBlobKey function?

BipBopBoop replied:
Yes, if this is being used multiple times, it's generally better to only initialise it once.

BipBopBoop commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -160,8 +167,18 @@ private BlobKey getBlobKey(HttpServletRequest request, String formInputElementNa
+    final BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    long size = blobInfo.getSize();
+    String type = blobInfo.getContentType();
+    if (size > 0 && FILETYPES.contains(type)) {
```
Is it useful to consider a max size as well? Do you want a user to be able to upload an image that is a terrabyte? Also is it possible to check the validity of the file before the user tries to upload it?

anna-giang replied:
Yep that would be better. I've put a limit of 5MB on it 👍

BipBopBoop replied:
Thanks :)

BipBopBoop reviewed (commented):
LGTM, just had a question about upload checks.

anna-giang committed:
Add upload filesize limit of 5MB

Prevents users from uploading files that are too large, which datastore/blobstore may not be able to handle.

anna-giang committed:
Add 'bytes' into comment regarding upload filesize limit

BipBopBoop commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -48,7 +48,7 @@
   private final HashSet<String> FILETYPES = new HashSet<String>(
       Arrays.asList("image/jpeg", "image/jpg", "image/png"));
-  // upload file size limit
+  // upload file size limit (bytes)
```
This is a good addition!

BipBopBoop reviewed (approved):
Thanks!
